resolwe/storage/models.py

```diff
@@ -92,6 +97,201 @@ def get_queryset(self) -> models.QuerySet:
+            )
+            for connector_name in connectors
+        ]
+
+        min_other_copies = rules.get("min_other_copies", 1)
+        default_delay = self._preprocess_delay(rules.get("delay"))
+
+        return (
+            FileStorage.objects.filter(
+                data__status=Data.STATUS_DONE,
+                created__lt=now()  # type: ignore
+                - models.Case(
+                    *whens,
+                    default=timedelta(days=default_delay),
+                    output_field=models.DurationField(),
+                ),
+            )
+            .annotate(
+                all_locations_count=models.Count(
+                    "storage_locations",
+                    filter=models.Q(
+                        storage_locations__status=StorageLocation.STATUS_DONE
+                    ),
+                    distinct=True,
+                )
+            )
+            .annotate(
+                open_logs=models.Exists(
+                    AccessLog.objects.filter(
+                        storage_location__file_storage=models.OuterRef("id"),
+                        storage_location__connector_name=connector_name,
+                        finished__isnull=True,
+                    )
+                )
+            )
+            .annotate(
+                accessed_lately=models.Exists(
+                    StorageLocation.objects.filter(
+                        file_storage=models.OuterRef("id"),
+                        connector_name=connector_name,
+                        last_update__gt=now()  # type: ignore
+                        - models.Case(
+                            *whens,
+                            default=timedelta(days=default_delay),
+                            output_field=models.DurationField(),
+                        ),
+                    )
+                )
+            )
+            .annotate(
+                default_connector=models.functions.Coalesce(
+                    models.Subquery(
+                        StorageLocation.objects.filter(
+                            file_storage=models.OuterRef("id"),
+                            status=StorageLocation.STATUS_DONE,
+                        )
+                        .annotate(
+                            priority=models.Case(
+                                *connectors_priority_whens,
+                                default=DEFAULT_CONNECTOR_PRIORITY,
+                                output_field=models.IntegerField(),
+                            )
+                        )
+                        .order_by("priority")
+                        .values_list("connector_name", flat=True)[:1]
+                    ),
+                    models.Value(""),
+                )
+            )
+            .filter(
+                all_locations_count__gt=min_other_copies,
+                open_logs=False,
+                accessed_lately=False,
+            )
+            .filter(
```

Comment:
Is there a reason for two separate filters?

Reply:
Just that I prefer to group logically connected (at least in my mind) filters together. I have merged them into one now.